cpp/src/io/parquet/reader_impl.cu

```diff
@@ -1162,15 +1162,23 @@ rmm::device_buffer reader::impl::decompress_page_data(
   for (const auto& codec : codecs) {
     if (codec.num_pages == 0) { continue; }
 
-    for_each_codec_page(codec.compression_type, [&](size_t page) {
-      auto dst_base = static_cast<uint8_t*>(decomp_pages.data());
-      comp_in.emplace_back(pages[page].page_data,
-                           static_cast<size_t>(pages[page].compressed_page_size));
-      comp_out.emplace_back(dst_base + decomp_offset,
-                            static_cast<size_t>(pages[page].uncompressed_page_size));
-
-      pages[page].page_data = static_cast<uint8_t*>(comp_out.back().data());
-      decomp_offset += comp_out.back().size();
+    for_each_codec_page(codec.compression_type, [&](size_t page_idx) {
+      auto dst_base = static_cast<uint8_t*>(decomp_pages.data()) + decomp_offset;
+      auto& page    = pages[page_idx];
+      // offset will only be non-zero for V2 pages
+      auto offset = page.def_lvl_bytes + page.rep_lvl_bytes;
+      // for V2 need to copy def and rep level info into place, and then offset the
+      // input and output buffers. otherwise we'd have to keep both the compressed
+      // and decompressed data.
+      if (offset) {
+        thrust::copy(rmm::exec_policy(_stream), page.page_data, page.page_data + offset, dst_base);
```

Comment:
Should this be `cudaMemcpyAsync` instead? I believe thrust is using the default stream and will force a sync.

Reply:
I was wondering if there was a better way to do this memcpy.  I'll give it a go.  Thanks!

Reply:
This might be wrong and thrust may be using cudaMemcpyAsync under the hood, but I don't know how it would select the correct stream.

Reply:
thrust is like ogres...layers!  I kind of like using the memcpy since I (think) I know exactly what it's doing.

Reply:
do you think it would be better to use a kernel like `gpu_copy_uncompressed_blocks` in gpuinflate.cu, rather than a bunch of sequential memcpy's?

Reply:
Trying to answer my own question...if I save up spans of data to copy, and then use `gpu_copy_uncompressed_blocks()` after the decompression is done, then execution time for the copy goes from around 68ms to 72us.  This is for a 3GB file with around 9250 pages.  Total execution time for `read_parquet()` is around 2.6s, and total execution time for a read/write with snappy compression is around 14.5s.  So is the increased complexity worth 3 orders of magnitude improvement in a section of code that accounts for at worst 0.4% of the total execution time?

Reply:
I wouldn't consider it amazingly more complex, but I'm not sure about the implementation change so it could be. It is hard for me to walk away from orders of magnitude of speed improvement. I understand your point though because it is a tradeoff and a small portion of the total time.

Reply:
I pushed the changes I was talking about, kept in ifdefs for now.  It's not *that* complex, but does add code to maintain.  I do like using the kernel since it's stupid fast. 😄

Reply:
I don't mind the extra complexity here. It's not unmaintainable in any sense, and seems to be the right way to do this.
My vote is to remove the ifdef and only keep the fast path.

Reply:
Sounds good. I think this is ready for review now.